T23800

tchx84 opened the pull request:
https://phabricator.endlessm.com/T23800

tchx84 committed:
game-state-service: initial version

https://phabricator.endlessm.com/T23800

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+    _dbus_error_name = 'com.endlessm.GameStateService.NotFoundException'
+
+
+class GameState(GObject.Object):
```
I think this can just be a subclass of `object`?

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+    def _monitor_restart_cb(self):
+        self._monitor_restart_id = None
+        self._monitor.handler_unblock(self._monitor_changed_id)
+        return False
```
use `GLib.SOURCE_REMOVE` for readability

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+        try:
+            success, state_data, tag = state_file.load_contents(None)
+            self._state = json.loads(state_data.decode())
+        except:
```
I'm not a big fan of `except: pass`, can you narrow this down to `IOError` or whatever?

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+                False,
+                0,
+                None)
+        except:
```
I think you can just delete this `except: pass`

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+        except:
+            pass
+        finally:
+            self._monitor_restart_id = GLib.timeout_add(
```
Uh oh, why is this necessary? Are you avoiding some kind of race condition?

cosimoc replied:
I think there are two other aspects to this pattern:
- we should defer writing the new state file to an idle or a short timeout. That way, if someone calls `set()` in a tight loop, we only write once at the end
- since you use the changes-done hint to reload the state from file, there should be no need to stop the file monitor while you write. But in fact there should be no need to have a file monitor at all, since all the writes go through the DBus service and you already update the state upon a call to `set()`

ptomato replied:
I understood from this morning's discussion that the idea behind the file monitor is that it's easily hackable by editing the text file.

cosimoc replied:
For development or is that something that we'd expose to users?

ptomato replied:
In any case for development, and also as something that curious users could discover.

tchx84 replied:
> Are you avoiding some kind of race condition?

Yes. Otherwise, it would always catch it owns writes. We don't want that.

> we should defer writing the new state file to an idle or a short timeout

I considered this, but when mixed with the "just edit the file" feature, I though it could turned into a tricky game. But I can give it a try.

> But in fact there should be no need to have a file monitor at all

There is, the "just edit the file" feature, is something we want to easily play with scenarios, and interact with the game state.

> For development or is that something that we'd expose to users?

For development really, but like @ptomato said, it would be cool if they find it and hack it.

cosimoc replied:
To be honest I don't really buy the "curious user" as a reason to have this...
About development, if it makes the code much more fragile, I do wonder if it's worth it.
I believe that it would be with my suggestion you could avoid the timeout, but on the other hand, if this is a service, it should also quit after a period of inactivity (which it doesn't do right now), and re-activate itself and recreate the state when receiving a new DBus call. So a file monitor is not really worth it, because the service is not supposed to stay around for very long anyway...

ptomato replied:
We should make a decision one way or the other on the "curious user" use case before merging this.

cosimoc replied:
I don't think that's an interesting use case for now. Let's proceed without and we can revisit later if we want to bring it back.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+        return self._state[key]
+
+    def set(self, key, value):
+        self._state[key] = value
```
Is there no need to unpack the variant like in GJS?

tchx84 replied:
There is, .unpack. I am using it now, since I switched to GDBus API.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+        self._state_to_file()
+
+
+class GameStateService(dbus.service.Object):
```
Could you combine the two classes into one, or is there a good reason not to do that?

tchx84 replied:
It makes it a lot easier _for me_ to read and maintain it, specially now with the GDBus API which is more verbose.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+    name = dbus.service.BusName(DBUS_IFACE, session_bus)
+    service = GameStateService(session_bus, DBUS_PATH)
+
+    mainloop = GObject.MainLoop()
```
This is supposed to be `GLib.MainLoop`, I think! Assuming this works, is it some sort of pygobject override? Also, I looked at the documentation of `dbus.mainloop.glib.DBusGMainLoop` and I don't understand how it integrates with this.

tchx84 replied:
Well, I haven't seen any python dbus or GDBus (using python) example using something different from `GObject.MainLoop()`, but `GLib.MainLoop()` seems to work as well.

ptomato commented on meson.build:
```diff
@@ -0,0 +1,24 @@
+project('game-state-service')
```
Probably good to add `version: '0.0.0'` here or something

ptomato commented on meson.build:
```diff
@@ -0,0 +1,24 @@
+install_data(
+    'game-state-service',
+    install_dir: get_option('bindir')
+)
```
at the end of the Meson run it's useful to print out the detected value of `prefix`, `bindir`, and `session_bus_services_dir` for the logs.

ptomato reviewed (changes requested):
I'm a bit surprised why you used Python for this when everything else is written in GJS, but it's probably not worth spending time to rewrite it right now. I think in the future we want to avoid context switching as much as we can though.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+
+import dbus
+import dbus.service
+import dbus.mainloop.glib
```
Using dbus-glib and dbus-python is a bit of a red flag for me. We should use GIO and GDBus for all new services we build.

tchx84 replied:
Switched to GDBus now, hopefully I am using it right, because it seems a bit more verbose.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,123 @@
+
+    @dbus.service.method(DBUS_IFACE, in_signature='sv', out_signature='')
+    def Set(self, key, value):
+        return self._state.set(key, value)
```
What is the interface we're using? Is it documented anywhere? Do we not need a signal to notify when the state has changed?

tchx84 replied:
The iface definition is explicit now, since I am using GDBus.

I think we could have a signal to let clients know that the state changed (globally), but I rather not try to mimic GObject properties-like API (for specific keys) or something like that for now. Since we agreed this Set/Get API was enough for now.

tchx84 commented:
@cosimoc @ptomato addressed all comments now, two specific things I will need your feedback on:

* `changed` signal. To keep it simple, since I am not exposing keys as if they were individual properties, it basically multi-casts the changes to the state.

* `Get` method. To keep this service as flexible as possible, the Set/Get receive/send variants. The thing is, is easy to unpack a receiving Variant, but not the opposite operation (once the object is stored, we lost its signature info). I tried writing a simple guess_signature method, but that got bad early-reviews, so now I just added the signature to the `Get` method (since it should be known anyway). Another alternative would be to store the signature in the state file, but that puts some extra responsibility on the file to not be modified manually.

ptomato commented on examples.py:
```diff
@@ -0,0 +1,77 @@
+from gi.repository import Gio, GLib
```
I think pygobject requires you to specify a version for gi.repository modules nowadays, or you get a warning. `gi.require_version('GLib', '2.0')` (check this in other files as well)

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+        self._state_save_deferred_id = None
+        self._state_from_file()
+
+    def _state_get_file(self):
```
This doesn't need to be a method; make it a constant.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+        state_file = self._state_get_file()
+        try:
+            state_file.replace_contents(
+                bytes(state_data.encode()),
```
`bytes` is unnecessary here, `encode()` already returns a `bytes`.

tchx84 replied:
This is gone now, since I am only using JSON-GLIb API.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+        self._loop.quit()
+
+    def run(self):
+        self._loop.run()
```
It should be possible to use GApplication (not GtkApplication, but the GLib base class) and simplify the DBus registration code. In particular, GApplication has a built-in "inactivity timeout" mechanism that will make the service quit when inactive for a given amount of time. We want that for this service.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+        except GLib.Error:
+            traceback.print_exc()
+        finally:
+            self._state_save_deferred_id = None
```
This is also done in the only caller of `_state_to_file()`, so it's redundant.

I think it would be better to remove this one and keep the other one, so that the reset of `_state_save_deferred_id` is next to `SOURCE_REMOVE`.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+    _DBUS_KEY_MESSAGE = 'key %s does not exist'
+    _DBUS_SIGNATURE_ERROR = 'com.endlessm.GameStateService.SignatureError'
+    _DBUS_SIGNATURE_MESSAGE = 'signature %s does not match key %s'
+    _DBUS_XML = '''
```
It may be worth considering splitting this in a real XML file. If you do that, and use gdbus-codegen to generate skeletons for the interface, the code to register objects would get simpler.
Optional for this PR if it's too much churn though.

tchx84 replied:
Ill create a separate issue for this.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+            else:
+                invocation.return_value(GLib.Variant('(v)', (variant,)))
+
+    def _on_bus_acquired(self, connection, name):
```
You can probably make this less verbose by inheriting from `Gio.Application` and implementing the `do_dbus_register` and `do_dbus_unregister` virtual methods.

ptomato commented on game-state-service:
```diff
@@ -0,0 +1,157 @@
+    _DBUS_KEY_ERROR = 'com.endlessm.GameStateService.KeyError'
+    _DBUS_KEY_MESSAGE = 'key %s does not exist'
+    _DBUS_SIGNATURE_ERROR = 'com.endlessm.GameStateService.SignatureError'
+    _DBUS_SIGNATURE_MESSAGE = 'signature %s does not match key %s'
```
It's a bit strange to have these format strings up here away from the format parameters, when they're only used in one place anyway. I'll leave that up to you whether you want to change that or keep all the strings in the same place.

ptomato commented:
> @cosimoc @ptomato addressed all comments now, two specific things I will need your feedback on:
> 
> * `changed` signal. To keep it simple, since I am not exposing keys as if they were individual properties, it basically multi-casts the changes to the state.

Sounds OK.

> * `Get` method. To keep this service as flexible as possible, the Set/Get receive/send variants. The thing is, is easy to unpack a receiving Variant, but not the opposite operation (once the object is stored, we lost its signature info). I tried writing a simple guess_signature method, but that got bad early-reviews, so now I just added the signature to the `Get` method (since it should be known anyway). Another alternative would be to store the signature in the state file, but that puts some extra responsibility on the file to not be modified manually.

Also sounds OK.

cosimoc reviewed (changes requested):
Me and @tchx84 discussed some improvements to the DBus interface to avoid the need to specify a signature when calling `Get()`. Other than that, this looks great overall and I only have a few comments here.

tchx84 commented:
@ptomato @cosimoc Now using JSON-GLib and Gio.Application, I think we got a cleaner API and the service behavior we wanted.

```
Set (key, variant)
Get (key) => variant
```

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,144 @@
+        return Gio.Application.do_dbus_register(self, connection, path)
+
+    def do_dbus_unregister(self, connection, path):
+        Gio.Application.do_dbus_register(self, connection, path)
```
Should be `do_dbus_unregister()`

tchx84 replied:
U.u

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,144 @@
+
+    def do_dbus_unregister(self, connection, path):
+        Gio.Application.do_dbus_register(self, connection, path)
+        connection.unregister_object(self._dbus_id)
```
You should probably check for `self._dbus_id is not None` before calling into this.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,144 @@
+
+    def save(self):
+        # If there is no pending save, return.
+        if self._save_deferred_id is None:
```
This does not look right; I expect a direct call to `save()` to bypass any deferred work and save, not save only if there is already a deferred work pending.

tchx84 replied:
It was a cheap optimization, basically, if there no pending save it means there is really nothing new to save (so we don't write to disk for nothing).

Maybe I should just rename it to "save_if_ needed"? hah

cosimoc replied:
I see. You could rename the method to something like `flush()`.

cosimoc commented on game-state-service:
```diff
@@ -0,0 +1,144 @@
+        generator = Json.Generator()
+        generator.set_root(root)
+        generator.set_pretty(True)
+        generator.set_indent(2)
```
Can you set these at construction time?
